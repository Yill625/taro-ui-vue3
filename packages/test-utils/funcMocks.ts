import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks