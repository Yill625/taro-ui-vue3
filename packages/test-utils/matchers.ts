import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers