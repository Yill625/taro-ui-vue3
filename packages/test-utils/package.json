{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.2.0",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}