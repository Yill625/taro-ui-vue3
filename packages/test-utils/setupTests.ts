import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()