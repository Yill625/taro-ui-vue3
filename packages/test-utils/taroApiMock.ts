import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
