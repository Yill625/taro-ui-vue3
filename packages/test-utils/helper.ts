import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}