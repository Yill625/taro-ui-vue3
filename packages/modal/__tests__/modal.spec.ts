import { genMountFn } from '@taro-ui-vue3/test-utils/helper'
import { h } from 'vue'
import AtModal from '../index.vue'

const mountFn = genMountFn(AtModal)

describe('AtModal', () => {
  it('should render modal with slot content and match snapshot', () => {
    const wrapper = mountFn({}, {
      default: [h('view', { class: 'slot' })]
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render modal with header and action and match snapshot', () => {
    const wrapper = mountFn({
      title: 'title',
      content: 'content',
      cancelText: 'cancel',
      confirmText: 'confirm'
    })
    const headerEl = wrapper.find('.at-modal__header')
    const actionEl = wrapper.find('.at-modal__action')
    const contentEl = wrapper.find('.at-modal__content')

    expect(headerEl.exists()).toBeTruthy()
    expect(headerEl.find('text').text()).toEqual('title')

    expect(contentEl.exists()).toBeTruthy()
    expect(contentEl.find('text').text()).toEqual('content')

    expect(actionEl.exists()).toBeTruthy()
    expect(
      actionEl.find('button:nth-child(1)').text()
    ).toEqual('cancel')
    expect(
      actionEl.find('button:nth-child(2)').text()
    ).toEqual('confirm')

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render modal without header', () => {
    const wrapper = mountFn({
      content: 'content'
    })
    expect(
      wrapper.find('.at-modal__header').exists()
    ).toBeFalsy()
  })

  it('should render modal with single button', () => {
    const wrapper = mountFn({
      content: 'content',
      confirmText: 'confirm'
    })
    const buttons = wrapper.findAll('.at-modal__action button')
    expect(buttons.length).toEqual(1)
    expect(buttons[0].text()).toEqual('confirm')
  })

  it('should render h5 button style', async () => {
    const wrapper = mountFn({
      content: 'content',
      cancelText: 'cancel',
      confirmText: 'confirm'
    })

    const button = wrapper
      .find('.at-modal__action button:nth-child(2)')

    expect(button.attributes('style')).toEqual('margin-top: 0px;')
  })

  it('should render prop isOpened', async () => {
    const wrapper = mountFn({ isOpened: true })

    expect(
      wrapper
        .find('.at-modal--active')
        .exists()
    ).toBeTruthy()

    await wrapper.setProps({
      content: 'content',
      cancelText: 'cancel',
      confirmText: 'confirm'
    })

    expect(
      wrapper
        .find('.at-modal--active')
        .exists()
    ).toBeTruthy()
  })
})

describe('AtModal events ', () => {
  it('should emit close, cancel and click events', async () => {
    const onCancel = jest.fn()
    const onConfirm = jest.fn()
    const onClose = jest.fn()
    const wrapper = mountFn({
      title: 'title',
      content: 'content',
      cancelText: 'cancel',
      confirmText: 'confirm',
      onClose,
      onCancel,
      onConfirm
    })

    const overlayEl = wrapper.find('.at-modal__overlay')
    const confirmBtnEl = wrapper
      .find('.at-modal__action button:last-child')
    const cancelBtnEl = wrapper
      .find('.at-modal__action button:first-child')

    await overlayEl.trigger('tap')
    await cancelBtnEl.trigger('tap')
    await confirmBtnEl.trigger('tap')

    expect(wrapper.emitted()).toHaveProperty('close')
    expect(wrapper.emitted()).toHaveProperty('cancel')
    expect(wrapper.emitted()).toHaveProperty('confirm')
    expect(onClose).toBeCalled()
    expect(onCancel).toBeCalled()
    expect(onConfirm).toBeCalled()
  })

  it('should not emit close if closeOnClickOverlay is false', async () => {
    const onClose = jest.fn()
    const wrapper = mountFn({
      title: 'title',
      content: 'content',
      cancelText: 'cancel',
      confirmText: 'confirm',
      closeOnClickOverlay: false,
      onClose
    })

    await wrapper.find('.at-modal__overlay').trigger('tap')
    expect(onClose).not.toBeCalled()
    expect(wrapper.emitted()).not.toHaveProperty('close')
  })

  it('should stop propagation of touch event', async () => {
    const stopPropagation = jest.fn()

    const wrapper = mountFn({}, {
      default: [h('view', { class: 'slot' })]
    })

    await wrapper
      .find('.at-modal')
      .trigger('touchmove', { stopPropagation })
    expect(
      wrapper.find('.at-modal').attributes('catchmove')
    ).toEqual('true')
    expect(stopPropagation).toBeCalled()

    await wrapper.setProps({
      title: 'title',
      content: 'content',
      cancelText: 'cancel',
      confirmText: 'confirm'
    })

    await wrapper
      .find('.at-modal')
      .trigger('touchmove', { stopPropagation })
    expect(
      wrapper.find('.at-modal').attributes('catchmove')
    ).toEqual('true')
    expect(stopPropagation).toBeCalled()
  })

  it('should render disableScroll in alipay', async () => {
    process.env.TARO_ENV = 'alipay'
    const wrapper = mountFn({ class: 'test' }, {
      default: [h('view', { class: 'slot' })]
    })

    expect(
      wrapper.find('.at-modal').attributes('disablescroll')
    ).toEqual('true')

    await wrapper.setProps({
      title: 'title',
      content: 'content',
      cancelText: 'cancel',
      confirmText: 'confirm'
    })

    expect(
      wrapper.find('.at-modal').attributes('disablescroll')
    ).toEqual('true')

    process.env.TARO_ENV = 'h5'
  })

})
