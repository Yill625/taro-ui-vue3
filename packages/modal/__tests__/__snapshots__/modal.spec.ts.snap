// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtModal should render modal with header and action and match snapshot 1`] = `
<view
  catchmove="true"
  class="at-modal"
>
  <view
    class="at-modal__overlay"
  />
  <view
    class="at-modal__container"
  >
    <view
      class="at-modal__header"
    >
      
      <text>
        title
      </text>
      
    </view>
    <scroll-view
      class="at-modal__content"
      scrolly="true"
    >
      
      <view
        class="content-simple"
      >
        <text>
          content
        </text>
      </view>
      
    </scroll-view>
    <view
      class="at-modal__footer at-modal__footer--simple"
    >
      <view
        class="at-modal__action"
      >
        
        <button>
          cancel
        </button>
        <button
          style="margin-top: 0px;"
        >
          confirm
        </button>
        
      </view>
    </view>
  </view>
</view>
`;

exports[`AtModal should render modal with slot content and match snapshot 1`] = `
<view
  catchmove="true"
  class="at-modal"
>
  <view
    class="at-modal__overlay"
  />
  <view
    class="at-modal__container"
  >
    
    <view
      class="slot"
    />
    
  </view>
</view>
`;
